src/algebraic_geometry/morphisms/quasi_separated.lean

```diff
@@ -274,4 +280,205 @@ begin
   { apply algebraic_geometry.quasi_separated_of_mono }
 end
 
+lemma exists_eq_pow_mul_of_is_affine_open (X : Scheme) (U : opens X.carrier) (hU : is_affine_open U)
+  (f : X.presheaf.obj (op U)) (x : X.presheaf.obj (op $ X.basic_open f)) :
+  ∃ (n : ℕ) (y : X.presheaf.obj (op U)),
+    y |_ X.basic_open f = (f |_ X.basic_open f) ^ n * x :=
+begin
+  have := (is_localization_basic_open hU f).2,
+  obtain ⟨⟨y, _, n, rfl⟩, d⟩ := this x,
+  use [n, y],
+  delta Top.presheaf.restrict_open Top.presheaf.restrict,
+  simpa [mul_comm x] using d.symm,
+end
+
+lemma exists_eq_pow_mul_of_is_compact_of_quasi_separated_space_aux (X : Scheme)
+  (S : X.affine_opens) (U₁ U₂ : opens X.carrier)
+  {n₁ n₂ : ℕ} {y₁ : X.presheaf.obj (op U₁)}
+  {y₂ : X.presheaf.obj (op U₂)} {f : X.presheaf.obj (op $ U₁ ⊔ U₂)}
+  {x : X.presheaf.obj (op $ X.basic_open f)}
+  (h₁ : S.1 ≤ U₁) (h₂ : S.1 ≤ U₂)
+  (e₁ : X.presheaf.map (hom_of_le $ X.basic_open_le
+    (X.presheaf.map (hom_of_le le_sup_left).op f) : _ ⟶ U₁).op y₁ =
+      X.presheaf.map (hom_of_le (by { erw X.basic_open_res, exact inf_le_left })).op
+        (X.presheaf.map (hom_of_le le_sup_left).op f) ^ n₁ *
+      (X.presheaf.map (hom_of_le (by { erw X.basic_open_res, exact inf_le_right })).op) x)
+  (e₂ : X.presheaf.map (hom_of_le $ X.basic_open_le
+    (X.presheaf.map (hom_of_le le_sup_right).op f) : _ ⟶ U₂).op y₂ =
+      X.presheaf.map (hom_of_le (by { rw X.basic_open_res, exact inf_le_left })).op
+        (X.presheaf.map (hom_of_le le_sup_right).op f) ^ n₂ *
+      (X.presheaf.map (hom_of_le (by { rw X.basic_open_res, exact inf_le_right })).op) x) :
+  ∃ n : ℕ, X.presheaf.map (hom_of_le $ h₁).op
+    ((X.presheaf.map (hom_of_le le_sup_left).op f) ^ (n + n₂) * y₁) =
+    X.presheaf.map (hom_of_le $ h₂).op
+      ((X.presheaf.map (hom_of_le le_sup_right).op f) ^ (n + n₁) * y₂) :=
+begin
+  have := (is_localization_basic_open S.2
+    (X.presheaf.map (hom_of_le $ le_trans h₁ le_sup_left).op f)),
+  obtain ⟨⟨_, n, rfl⟩, e⟩ :=
+    (@is_localization.eq_iff_exists _ _ _ _ _ _ this (X.presheaf.map (hom_of_le $ h₁).op
+      ((X.presheaf.map (hom_of_le le_sup_left).op f) ^ n₂ * y₁))
+    (X.presheaf.map (hom_of_le $ h₂).op
+      ((X.presheaf.map (hom_of_le le_sup_right).op f) ^ n₁ * y₂))).mp _,
+  swap,
+  { simp only [map_pow, ring_hom.algebra_map_to_algebra, map_mul, ← comp_apply,
+      ← functor.map_comp, ← op_comp] at ⊢,
+    transitivity X.presheaf.map (hom_of_le _).op f ^ n₂ *
+      X.presheaf.map (hom_of_le _).op f ^ n₁ * X.presheaf.map (hom_of_le _).op x,
```

Comment:
It is nice to have Lean create the goals, but it is a bit scary to see goals like "has_mul ..." pop. It seems if you replace some of `_` by a term which is an explicit inequality between opens, these irrelevant goals will disappear.